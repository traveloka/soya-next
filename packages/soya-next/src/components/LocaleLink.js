import React from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import { compose } from 'redux';
import { format, resolve } from 'url';
import { withRouter } from 'next/router';
import withLocale from '../i18n/withLocaleComponent';
import { localeShape } from '../constants/PropTypes';
import { toPath } from '../utils/locale';

class LocaleLink extends React.Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    defaultLocale: PropTypes.string.isRequired,
    locale: localeShape.isRequired,
    router: PropTypes.object,
  };

  render() {
    const {
      children,
      defaultLocale,
      locale,
      router,
      ...props
    } = this.props;
    let { as, href } = props;
    delete props.siteLocales;

    const { language, country } = locale;
    const localeSegment = toPath(locale, defaultLocale);

    href = href ? format(href) : '';
    as = as ? format(as) : '';
    const sep = href.indexOf('?') === -1 ? '?' : '&';

    return (
      <Link
        {...props}
        as={localeSegment + resolve(router.pathname, as || href)}
        href={`${resolve(router.pathname, href)}${sep}locale=${language}-${country}`}
      >
        {children}
      </Link>
    );
  }
}

export default compose(
  withLocale,
  withRouter,
)(LocaleLink);
