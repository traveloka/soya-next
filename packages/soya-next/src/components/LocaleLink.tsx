import React from "react";
import { compose } from "redux";
import Link, { LinkProps } from "next/link";
import { withRouter } from "next/router";
import withLocale, {
  WithLocaleComponentInjectedProps,
} from "../i18n/withLocaleComponent";
import { toPath } from "../utils/locale";
import { convertLegacyUrlObjectToWhatwgUrl, resolveUrl } from "../utils/url";
import type { WithRouterProps } from "next/dist/client/with-router";

export interface LocaleLinkInjectedProps
  extends LinkProps,
    WithLocaleComponentInjectedProps,
    WithRouterProps {
  children: React.ReactNode;
}
export interface LocaleLinkProps extends LinkProps {
  locale?: WithLocaleComponentInjectedProps["locale"];
}

/**
 * Soya Next - Locale Link
 * ---
 *
 * A locale aware `<Link />` component for routing.
 *
 * **Note**: Please ensure that `<LocaleLink />` is rendered
 * within `createPage` hierarchy.
 */
export class LocaleLink extends React.Component<LocaleLinkInjectedProps> {
  getAs(asPath: NonNullable<LinkProps["as"]>) {
    if (typeof asPath === "string") {
      return asPath;
    }

    return convertLegacyUrlObjectToWhatwgUrl(asPath).pathname;
  }

  getHref(href: NonNullable<LinkProps["href"]>) {
    if (typeof href === "string") {
      return href;
    }

    return convertLegacyUrlObjectToWhatwgUrl(href).pathname;
  }

  render() {
    const { children, defaultLocale, locale, router, siteLocales, ...props } =
      this.props;
    let { as, href } = props;

    const { language, country } = locale || {};
    const localeSegment = toPath(locale, defaultLocale);

    href = href ? this.getHref(href) : "";
    as = as ? this.getAs(as) : "";
    const sep = href.indexOf("?") === -1 ? "?" : "&";

    return (
      <Link
        {...props}
        as={localeSegment + resolveUrl(router.pathname, as || href)}
        href={`${resolveUrl(
          router.pathname,
          href
        )}${sep}locale=${language}-${country}`}
      >
        {children}
      </Link>
    );
  }
}

/**
 * Soya Next - Locale Link
 * ---
 * A locale aware `<Link />` component for routing.
 *
 * **Note**: Please ensure that `<LocaleLink />` is rendered
 * within `createPage` hierarchy.
 */
export default compose<React.ComponentType<LocaleLinkProps>>(
  withLocale,
  withRouter
)(LocaleLink);
