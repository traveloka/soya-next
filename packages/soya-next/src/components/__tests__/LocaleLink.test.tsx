import React from "react";
import { render } from "@testing-library/react";
import { LocaleLink } from "../LocaleLink";

// MOCK <Link /> component for shallow rendering.
jest.mock("next/link", () => {
  return jest.fn(props => {
    return <span {...props}>{props.children}</span>;
  });
});

describe("<LocaleLink />", () => {
  const context = {
    locale: {
      country: "id",
      language: "id",
    },
    defaultLocale: "id-id",
    siteLocales: ["id-id", "en-id"],
    router: {
      pathname: "/",
    },
  };

  it("should render <Link /> to navigate from default locale", () => {
    const { container } = render(
      <LocaleLink {...(context as any)} href="/about">
        <a>{"Tentang"}</a>
      </LocaleLink>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render <Link /> to navigate from non-default locale", () => {
    const locale = {
      language: "en",
      country: "id",
    };
    const { container } = render(
      <LocaleLink {...(context as any)} href="/about" locale={locale}>
        <a>{"About"}</a>
      </LocaleLink>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render <Link /> to navigate to custom route", () => {
    const { container } = render(
      <LocaleLink
        {...(context as any)}
        as="p/hello-world"
        href="/post?title=Hello World"
      >
        <a>{"Hello World"}</a>
      </LocaleLink>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render <Link /> to navigate to another locale", () => {
    const locale = {
      language: "en",
      country: "id",
    };
    const { container } = render(
      <LocaleLink {...(context as any)} href={"/"} locale={locale}>
        <a>{"English (Indonesia)"}</a>
      </LocaleLink>
    );
    expect(container).toMatchSnapshot();
  });

  it("should accept locale object", () => {
    const { container } = render(
      <LocaleLink
        {...(context as any)}
        locale={{ language: "en", country: "sg" }}
        href="/about"
      >
        <a>{"About"}</a>
      </LocaleLink>
    );
    expect(container).toMatchSnapshot();
  });

  it("should accept href and 'as' as object", () => {
    const locale = {
      language: "en",
      country: "id",
    };
    const { container } = render(
      <LocaleLink
        {...(context as any)}
        locale={locale}
        as={{ pathname: "/blog/my-post" }}
        href={{
          pathname: "/blog/[slug]",
          query: { slug: "my-post" },
        }}
      >
        <a>{"My Post"}</a>
      </LocaleLink>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render the <Link /> successfully while passing wrong as or href path", () => {
    const { container } = render(
      <LocaleLink locale={null} href={null} {...(context as any)}>
        <a>{"My Post"}</a>
      </LocaleLink>
    );
    expect(container).toMatchSnapshot();
  });
});
