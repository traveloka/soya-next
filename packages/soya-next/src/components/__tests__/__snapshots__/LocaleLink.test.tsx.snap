// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocaleLink /> should accept href and 'as' as object 1`] = `
<div>
  <span
    as="/en/blog/my-post"
    href="/blog/[slug]?locale=en-id"
  >
    <a>
      My Post
    </a>
  </span>
</div>
`;

exports[`<LocaleLink /> should accept locale object 1`] = `
<div>
  <span
    as="/en-sg/about"
    href="/about?locale=en-sg"
  >
    <a>
      About
    </a>
  </span>
</div>
`;

exports[`<LocaleLink /> should render <Link /> to navigate from default locale 1`] = `
<div>
  <span
    as="/about"
    href="/about?locale=id-id"
  >
    <a>
      Tentang
    </a>
  </span>
</div>
`;

exports[`<LocaleLink /> should render <Link /> to navigate from non-default locale 1`] = `
<div>
  <span
    as="/en/about"
    href="/about?locale=en-id"
  >
    <a>
      About
    </a>
  </span>
</div>
`;

exports[`<LocaleLink /> should render <Link /> to navigate to another locale 1`] = `
<div>
  <span
    as="/en/"
    href="/?locale=en-id"
  >
    <a>
      English (Indonesia)
    </a>
  </span>
</div>
`;

exports[`<LocaleLink /> should render <Link /> to navigate to custom route 1`] = `
<div>
  <span
    as="/p/hello-world"
    href="/post?title=Hello%20World&locale=id-id"
  >
    <a>
      Hello World
    </a>
  </span>
</div>
`;

exports[`<LocaleLink /> should render the <Link /> successfully while passing wrong as or href path 1`] = `
<div>
  <span
    as="/"
    href="/?locale=id-id"
  >
    <a>
      My Post
    </a>
  </span>
</div>
`;
